File: 0721-accounts-merge/0721-accounts-merge.cpp
class Solution {
private:
    int find(vector<int>& parent, int u) {
        if (parent[u] == u) {
            return u;
        }

        parent[u] = find(parent, parent[u]);
        return parent[u];
    }

public:
    vector<vector<string>> accountsMerge(vector<vector<string>>& accounts) {
        vector<vector<string>>::size_type n = accounts.size();
        unordered_map<string, int> emailToId{};
        int emailId = 0;

        for (vector<string>& acc : accounts) {
            vector<string>::size_type accSize = acc.size();

            for (int i = 1; i < accSize; ++i) {
                auto it = emailToId.find(acc[i]);

                if (it == emailToId.end()) {
                    emailToId.insert({acc[i], emailId++});
                }
            }
        }

        vector<int> parent(emailId);
        std::iota(parent.begin(), parent.end(), 0);

        int resSize = emailId;

        for (vector<string>& acc : accounts) {
            vector<string>::size_type accSize = acc.size();
            int firstRep = find(parent, emailToId[acc[1]]);

            for (int i = 2; i < accSize; ++i) {
                int id = emailToId[acc[i]];
                int rep = find(parent, id);

                if (firstRep != rep) {
                    --resSize;
                    parent[rep] = firstRep;
                }
            }
        }

        vector<set<string>> temp(resSize, set<string>{});
        unordered_map<int, int> repToIdx{};
        vector<string> idxToName(resSize);
        int idx = 0;

        for (vector<string> const& acc : accounts) {
            vector<string>::size_type accSize = acc.size();
            int rep = find(parent, emailToId[acc[1]]);

            auto it = repToIdx.find(rep);

            if (it == repToIdx.end()) {
                repToIdx.insert({rep, idx++});
            }

            int idx = repToIdx[rep];
            set<string>& newAcc = temp[idx];
            idxToName[idx] = acc[0];

            for (int i = 1; i < accSize; ++i) {
                newAcc.insert(acc[i]);
            }
        }

        vector<vector<string>> result{};
        result.reserve(resSize);

        for (int i = 0; i < resSize; ++i) {
            result.push_back(vector<string>{idxToName[i]});
            
            for (string const& email : temp[i]) {
                result[i].push_back(email);
            }
        }

        return result;
    }
};

File: 0721-accounts-merge/0721-accounts-merge.cs
public class Solution {
    public IList<IList<string>> AccountsMerge(IList<IList<string>> accounts) {
        Dictionary<string, int> emailToId = new Dictionary<string, int>();
        Dictionary<int, string> emailIdToName = new Dictionary<int, string>();
        Dictionary<int, int> parent = new Dictionary<int, int>();
        int length = accounts.Count;
        int emailId = 0;

        for (int i = 0; i < length; ++i) {
            IList<string> account = accounts[i];
            int accLength = account.Count;

            for (int j = 1; j < accLength; ++j) {
                string email = account[j];

                if (emailToId.TryAdd(email, emailId)) {
                    ++emailId;
                }

                int id = emailToId[email];
                parent[id] = emailToId[account[1]];
                emailIdToName[id] = account[0];
            }
        }

        int find(Dictionary<int, int> parent, int id) {
            if (parent[id] == id) {
                return id;
            }

            parent[id] = find(parent, parent[id]);
            return parent[id];
        }

        for (int i = 0; i < length; ++i) {
            IList<string> account = accounts[i];
            int accLength = account.Count;
            int firstRep = find(parent, emailToId[account[1]]);

            for (int j = 2; j < accLength; ++j) {
                int rep = find(parent, emailToId[account[j]]);
                
                if (rep != firstRep) {
                    parent[rep] = firstRep;
                }
            }
        }

        Dictionary<int, List<string>> repToEmails = new Dictionary<int, List<string>>();

        foreach (KeyValuePair<string, int> emailIdPair in emailToId) {
            int rep = find(parent, emailIdPair.Value);
            
            if (!repToEmails.ContainsKey(rep)) {
                repToEmails[rep] = new List<string>();
            }

            repToEmails[rep].Add(emailIdPair.Key);
        }

        foreach(List<string> emails in repToEmails.Values) {
            emails.Sort(StringComparer.Ordinal);
        }

        IList<IList<string>> result  = new List<IList<string>>();

        foreach (List<string> emails in repToEmails.Values) {
            string name = emailIdToName[emailToId[emails[0]]];
            result.Add(new List<string>());
            
            result.Last().Add(name);
            foreach (string email in emails) {
                result.Last().Add(email);
            }
        }

        return result;
    }
}